Dynamic Memory Limits for LST Objects

GNiendorf opened the pull request:
### Dynamic Memory Limits

I compute a dynamic max for each module that is the max possible number of objects that can be created given the number of lower level objects (so for example the number of possible T5's for a given module given the total number of T3's in modules where two T3's can be connected). This upper bound provides a large reduction in the occupancies. This is also accomplished with no new truncations, since this dynamic max represents the maximum number of possible objects that can be created.

For the segments, I had to rearrange the addPixelSegmentToEvent() function so that the segments container is instantiated after the MDs container is filled. Right now the segments object is instantiated during the pLS loading, which is done before the MDs container is filled.

### Object Size Reduction

Reduction in object sizes (total occupancy basically) at current max occupancies (averaged over 10 events, 0.8 GeV):
- 682,404 -> 303,891 (55% decrease) for MD's.
- 1,930,994->857,880 (56% decrease) for Segments.
- 1,980,235 -> 737,502 (63% decrease) for T3's.
- 502,053 -> 212,953 (58% decrease) for T5's.

### Total Memory Decrease

For single stream, 100 events at current occupancies:

1695MiB -> 1275MiB (**25%** decrease, 0.8 GeV)

GNiendorf commented:
Performance plots look good, and negligible changes in timing. I moved the location of the pixel segments kernel, but I may fix this soon. So the timing for the Hits category went down and the timing for the LS (where I threw the kernel) goes up, but total timing is the same. This PR is ready for review though, everything works.

slava77 commented:
/run all

GNiendorf commented:
Should I rebase to master?

slava77 commented:
> Should I rebase to master?

I don't think it will help by itself; the CI is not pulling in the right version of alpaka.
... I guess if the target branch is not master but instead something older than Jan 21, the tests may work

GNiendorf commented:
/run all

slava77 commented:
> /run all

this almost worked with switching to `some_branch`; the CMSSW part failed due to a missing `HeterogeneousCore/AlpakaMath` I thought that addpkg was not throwing.
A better guess in this context is to use `CMSSW_14_2_0_pre4_workflowsAndGeneralFunctions_squashed` to run the tests ... although that's a Dec 4 master and `dynamic_occ` is from Dec 16 and can pull in some extra stuff.
Still, it's a hack and it's better to get the CI fixed.

ariostas commented:
I changed the target branch because I thought that the CI should work now that it uses the latest IB release, but I just noticed that it failed for another PR. I'll look into it.

ariostas commented:
/run cmssw

ariostas commented:
I'll change the CI so that it does the merge before checking out the packages

slava77 commented:
> There's going to have to be some (hopefully minor) rebasing now that my PR was merged into CMSSW. I'll update our master to see if there are conflicts.

5 files with conflicts now

GNiendorf commented:
Should I just rebase to master and fix the conflicts? Seems like the CMSSW run failed for unrelated reasons.

GNiendorf commented:
Just squashed the commits, will rebase now to fix the conflicts if that is fine.

slava77 commented:
> Should I just rebase to master and fix the conflicts? Seems like the CMSSW run failed for reasons unrelated to this PR.

I'd rebase (to master or to SegmentLinking:ariostas/integration_pr_followups should work)

GNiendorf commented:
/run

GNiendorf commented:
/run all

GNiendorf commented:
I'm not sure if this will run. When I try to run this locally I get errors from @ariostas latest PR, but I'm not sure if it's just something in my local setup.

../../src/alpaka/Hit.h(107): error: namespace "alpaka_std" has no member "lower_bound"
          auto found_pointer = alpaka_std::lower_bound(modules.mapdetId(), modules.mapdetId() + nModules, iDetId);
                                           ^

../../src/alpaka/Hit.h(108): error: no instance of function template "std::distance" matches the argument list
            argument types are: (const unsigned int *, <error-type>)
          int found_index = std::distance(modules.mapdetId(), found_pointer);

GNiendorf commented:
Looks like I missed something when rebasing, I will fix that now.

ariostas commented:
@GNiendorf You'll need to checkout `HeterogeneousCore/AlpakaInterface` since the release you're using doesn't have that function.

GNiendorf commented:
/run all

GNiendorf commented:
@ariostas The tests should still run though, correct?

ariostas commented:
> The tests should still run though, correct?

Yeah, let's see what happens

GNiendorf commented:
/run cmssw

GNiendorf commented:
@slava77 Do you have any comments for this PR, or should I submit it to CMSSW?

slava77 commented:
> @slava77 Do you have any comments for this PR, or should I submit it to CMSSW?

- I missed the part when `PixelSegmentSoA` was added and will need some time to check the code. 
- I'm not sure I follow the comment in the PR description `It looks like the process of moving the segments over to this method resulted in a net increase in memory usage`, while above that the segments are stated to decrease. 
- Please review the description to cover the above (and be self-consistent)

what's the plan with #147 ? I thought that you wanted that more or less together with this.
Does this PR interfere with changes in #146 ?

ariostas commented:
I went through the changes and they look good to me

GNiendorf commented:
> * I missed the part when `PixelSegmentSoA` was added and will need some time to check the code.
> * I'm not sure I follow the comment in the PR description `It looks like the process of moving the segments over to this method resulted in a net increase in memory usage`, while above that the segments are stated to decrease.

I moved some arrays (the pLS hit indices and their delta phi) onto the GPU in the pixelSegments SoA but corrected this with a later update. I removed the edit from the PR description and just list the current total memory savings for 0.8 GeV at 100 events to be self-consistent.

> what's the plan with #147 ? I thought that you wanted that more or less together with this. Does this PR interfere with changes in #146 ?

 I think it makes sense to merge this PR separately from #147. I think it's still unclear what the occupancies should be set to, and that will involve its own efficiency and memory changes. Therefore I think it's better to merge this in first before addressing the occupancy tuning in that PR.

slava77 commented on RecoTracker/LSTCore/src/alpaka/TrackCandidate.h:
```diff
@@ -128,8 +128,8 @@ namespace ALPAKA_ACCELERATOR_NAMESPACE::lst {
+          float eta2 = pixelSegments.eta()[pLS_jx - prefix];
+          float phi2 = pixelSegments.phi()[pLS_jx - prefix];
           float dEta = alpaka::math::abs(acc, (eta1 - eta2));
           float dPhi = calculate_dPhi(phi1, phi2);
```
answering my question about connection to #146  : this will be a conflicting line; and there is another one later.
Just as a heads up to @VourMa

slava77 commented on RecoTracker/LSTCore/src/alpaka/Segment.h:
`#include "RecoTracker/LSTCore/interface/PixelSegmentsSoA.h"` is missing

slava77 commented on RecoTracker/LSTCore/src/alpaka/LSTEvent.h:
```diff
@@ -106,26 +112,26 @@ namespace ALPAKA_ACCELERATOR_NAMESPACE::lst {
+                                 std::vector<char> const& isQuad);
 
     void createMiniDoublets();
+    void createPixelSegmentToEvent(std::vector<unsigned int> hitIndices0,
```
I'd call this `addPixelSegmentToEventPart2` or `addPixelSegmentToEventFinalize`
otherwise the naming is quite confusing ("add to event" and "create to event")

slava77 commented on RecoTracker/LSTCore/src/alpaka/LSTEvent.h:
```diff
@@ -106,26 +112,26 @@ namespace ALPAKA_ACCELERATOR_NAMESPACE::lst {
-                                std::vector<int> const& superbin,
-                                std::vector<PixelType> const& pixelType,
-                                std::vector<char> const& isQuad);
+    void loadPixelSegmentToEvent(std::vector<float> const& ptIn,
```
I'd call this `addPixelSegmentToEventPart1` or `addPixelSegmentToEventStart` or smth

slava77 commented on RecoTracker/LSTCore/src/alpaka/Triplet.h:
```diff
@@ -848,15 +849,28 @@ namespace ALPAKA_ACCELERATOR_NAMESPACE::lst {
+        int occupancy = alpaka::math::min(acc, dynamic_count, matrix_cap);
 
         ranges.tripletModuleOccupancy()[i] = occupancy;
         unsigned int nTotT = alpaka::atomicAdd(acc, &nTotalTriplets, occupancy, alpaka::hierarchy::Threads{});
```
ah, I guess `unsigned int` in the `Segment.h` arises from here.
To be stricter, unsigned for nX is better, but then perhaps the inputs should be unsigned as well (`nTotalTriplets` and `occupancy` are both signed)

GNiendorf replied:
I decided to leave all of left the signs the way they are now, since I think it would be better to do dedicated code cleanup like this in its own PR.

slava77 reviewed (changes requested):
this will also likely need a rebase very soon, once the upstream variant of #146 is merged

slava77 commented:
> For the segments, I had to rearrange the createPixelSegmentToEvent() function so that the segments object is created after the mini-doublets are created. Right now in the code the segments are created during the pLS loading, which is done before the MDs are created.

mostly for the purpose of the CMSSW PR, to have more clarity perhaps instead of "MDs are created" say "the MD container is instantiated". Our `create*` methods in most cases also run the algorithms and populate relevant containers.

slava77 commented on RecoTracker/LSTCore/src/alpaka/Kernels.h:
```diff
@@ -390,15 +390,15 @@ namespace ALPAKA_ACCELERATOR_NAMESPACE::lst {
           }
           if (secondpass) {
             float dEta = alpaka::math::abs(acc, eta_pix1 - eta_pix2);
             float dPhi = calculate_dPhi(phi_pix1, phi_pix2);
```
uhm, I've updated the target/master branch, but this line still has `calculate_dPhi`; odd.

GNiendorf replied:
I would have to rebase for it to change, wouldn't I?

slava77 replied:
well, yes and no. you had only one non-merge commit, it should still properly detect the diff.
True, if you rebase it wouldn't have a way to confuse.

GNiendorf commented:
/run all

GNiendorf commented:
/run all

slava77 commented:
@GNiendorf 
please run timing tests on GPU. The host-side per-object loops may show some significant slowdown

GNiendorf commented:
Thank you @ariostas!

slava77 commented:
> Here's a timing comparison on cgpu-1.

thank you; for the single-stream perhaps I should ignore the hits column
